src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.IO;
+using System.Management.Automation;
+using System.Management.Automation.Host;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Net.NetworkInformation;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.PowerShell
+{
+    /// <summary>
+    /// A Helper class for printing notification on PowerShell startup when there is a new update.
+    /// </summary>
+    internal static class UpdatesNotification
+    {
+        private const string UpdateCheckOptOutEnvVar = "POWERSHELL_UPDATECHECK_OPTOUT";
+        private const string Last3ReleasesUri = "https://api.github.com/repos/PowerShell/PowerShell/releases?per_page=3";
+        private const string LatestReleaseUri = "https://api.github.com/repos/PowerShell/PowerShell/releases/latest";
+
+        private const string SentinelFileName = "_sentinel_";
+        private const string DoneFileNameTemplate = "sentinel-{0}-{1}-{2}.done";
+        private const string DoneFileNamePattern = "sentinel-*.done";
+        private const string UpdateFileNameTemplate = "update_{0}_{1}";
+        private const string UpdateFileNamePattern = "update_v*.*.*_????-??-??";
+
+        private static readonly EnumerationOptions s_enumOptions = new EnumerationOptions();
+        private static readonly string s_cacheDirectory = Path.Combine(Platform.CacheDirectory, PSVersionInfo.GitCommitId);
+
+        /// <summary>
+        /// Gets a value indicating whether update notification should be done.
+        /// </summary>
+        internal static readonly bool CanNotifyUpdates = !Utils.GetOptOutEnvVariableAsBool(UpdateCheckOptOutEnvVar, defaultValue: false)
```

Comment:
Good question. I thought about this and decided to not do the check here because:
1. having the check here will only help the private builds to save a task creation;
2. for preview and stable versions that we really care, `PSVersionInfo.GitCommitId.Length > 40` is always false, so having the check here means doing unnecessary work at the type initialization phase, which happens at the startup time. With the current code, we do the check in the task, which runs after startup.

Reply:
Perhaps we could do CanNotifyUpdates in CheckForUpdates() after delay finished.

Reply:
Agreed